### PR TITLE
Database projects deploy

### PR DESCRIPTION
**Pull request by Benjin**
* Hooking up deployment of SQL projects to the project build functionality and database selection UI
* Adding ADS-side plumbing for sqlcmdvars

### CODE REVIEW

**Comment by Benjin** [comment]
No tests yet; those are next.  I wanted to get this out there so people can start taking a look while I add tests/make things testable.

SQLCMD vars are there on this side, but I haven't sent out the PR for the SQL Tools Service side yet.  When I used my own STS DLLs, variable replacement works (yay!), but I encountered an issue in project build related to it that I haven't had a chance to debug yet, so I reverted those since I wanted to focus on this PR.  SQLCMD vars existing on this side without the changes in STS isn't causing any breakage in dacpac deployment, so I figured I'd leave them it.

File: extensions/sql-database-projects/src/controllers/projectController.ts [hunk 1]
```diff
@@ -119,25 +122,89 @@ export class ProjectsController {
...
+		let result;
+
+		if (profile as IDeploymentProfile) {
+			result = await dacFxService.deployDacpac(dacpacPath, profile.databaseName, (<IDeploymentProfile>profile).upgradeExisting, profile.connectionUri, TaskExecutionMode.execute, profile.sqlCmdVariables);
+		}
+		else {
+			result = await dacFxService.generateDeployScript(dacpacPath, profile.databaseName, profile.connectionUri, TaskExecutionMode.execute, profile.sqlCmdVariables);
+		}
+
+		console.log(result);
```
Comment (kisantia) [inline extensions/sql-database-projects/src/controllers/projectController.ts:194]: remove console.log

File: extensions/sql-database-projects/src/dialogs/deployDatabaseDialog.ts [hunk 2]
```diff
@@ -104,14 +104,59 @@ export class DeployDatabaseDialog {
...
+				password: '', // TODO: secure password storage
+				authenticationType: dataSource.integratedSecurity ? 'Integrated' : 'SqlAuth',
+				savePassword: false,
+				providerName: 'MSSQL',
+				saveProfile: true,
+				id: dataSource.name + '-dataSource',
+				options: []
+			};
+
+			const connResult = await azdata.connection.connect(connProfile, false, false);
```
Comment (kisantia) [inline extensions/sql-database-projects/src/dialogs/deployDatabaseDialog.ts:128]: does the connection dialog need to pop up if it's SqlAuth?

File: extensions/sql-database-projects/src/dialogs/deployDatabaseDialog.ts [hunk 3]
```diff
@@ -104,14 +104,59 @@ export class DeployDatabaseDialog {
...
+			connId = connResult.connectionId;
+		}
+		else {
+			const connection = (await azdata.connection.getConnections()).find(c => c.connectionName === this.targetConnectionTextBox!.value)!;
```
Comment (kisantia) [inline extensions/sql-database-projects/src/dialogs/deployDatabaseDialog.ts:132]: Would this work if the connection isn't named? Also, the deployDatabaseDialog.connection is set to the selected connection when a connection is selected using the Edit button so that can be used instead of searching through the connections.

File: extensions/sql-database-projects/src/dialogs/deployDatabaseDialog.ts [hunk 4]
```diff
@@ -238,8 +283,8 @@ export class DeployDatabaseDialog {
 
 			// show connection name if there is one, otherwise show connection string
 			if (this.connection.options['connectionName']) {
-				this.targetConnectionTextBox!.value = this.connection.options['connectionName'];
 			} else {
+				this.targetConnectionTextBox!.value = this.connection.options['connectionName'];
```
Comment (kisantia) [inline extensions/sql-database-projects/src/dialogs/deployDatabaseDialog.ts:287]: why was this moved from the if?

File: extensions/sql-database-projects/src/controllers/projectController.ts [hunk 5]
```diff
@@ -119,25 +122,89 @@ export class ProjectsController {
...
 			argument: this.buildHelper.constructBuildArguments(project.projectFilePath, this.buildHelper.extensionBuildDirPath)
 		};
-		await this.netCoreTool.runDotnetCommand(options);
+		try {
+			await this.netCoreTool.runDotnetCommand(options);
+
+			return path.join(project.projectFolderPath, 'bin', 'Debug', `${project.projectFileName}.dacpac`);
+		}
+		catch (err) {
+			this.apiWrapper.showErrorMessage(`Build failed.  Check output pane for more details.  ${utils.getErrorMessage(err)}`);
```
[the remark this answers is not in the document]
Comment (Charles-Gagnon) [reply extensions/sql-database-projects/src/controllers/projectController.ts:154]: And make sure you localize anything that's shown to the user

File: extensions/sql-database-projects/src/dialogs/deployDatabaseDialog.ts (hunk 4 above)
Comment (Benjin) [reply extensions/sql-database-projects/src/dialogs/deployDatabaseDialog.ts:287]: oops, I don't think I meant to do this

File: extensions/sql-database-projects/src/dialogs/deployDatabaseDialog.ts (hunk 3 above)
Comment (Benjin) [reply extensions/sql-database-projects/src/dialogs/deployDatabaseDialog.ts:132]: Hmm... to both this comment and the one about SqlAuth, I think I'll need to some help defining the behavior here.  Ultimately, this method's output should be the Uri within azdata.connection that should be used for dacFxService.

File: extensions/sql-database-projects/src/dialogs/deployDatabaseDialog.ts (hunk 2 above)
Comment (Benjin) [reply extensions/sql-database-projects/src/dialogs/deployDatabaseDialog.ts:130]: I think for the purposes of this PR (stringing together pieces), it's okay to say that the SqlAuth credentials come from datasources.json, and it the connection will fail if it's missing password.  (see other comment about needing to define more concrete behavior here)

File: extensions/sql-database-projects/src/dialogs/deployDatabaseDialog.ts [hunk 6]
```diff
@@ -104,17 +106,68 @@ export class DeployDatabaseDialog {
...
+
+		if (this.dataSourcesRadioButton?.checked) {
+			const dataSource = (this.dataSourcesDropDown!.value! as DataSourceDropdownValue).dataSource;
+
+			const connProfile: azdata.IConnectionProfile = {
+				serverName: dataSource.server,
+				databaseName: dataSource.database,
+				connectionName: dataSource.name,
+				userName: '',
+				password: '', // TODO: secure password storage
```
Comment (Benjin) [inline extensions/sql-database-projects/src/dialogs/deployDatabaseDialog.ts:121]: add to datasources.json, and add a TODO note there about storing/accessing/prompting for creds

File: extensions/sql-database-projects/src/controllers/projectController.ts [hunk 7]
```diff
@@ -119,23 +122,77 @@ export class ProjectsController {
...
+	 */
+	public async buildProject(treeNode: BaseProjectTreeItem): Promise<string>;
+	/**
+	 * Builds a project, producing a dacpac
+	 * @param project Project to be built
+	 * @returns path of the built dacpac
+	 */
+	public async buildProject(project: Project): Promise<string>;
+	public async buildProject(context: Project | BaseProjectTreeItem): Promise<string | undefined> {
+		const project: Project = context instanceof Project ? context : this.getProjectContextFromTreeNode(context);
```
Comment (Benjin) [inline extensions/sql-database-projects/src/controllers/projectController.ts:138]: Consolidate

File: extensions/sql-database-projects/src/dialogs/deployDatabaseDialog.ts [hunk 8]
```diff
@@ -104,17 +106,68 @@ export class DeployDatabaseDialog {
...
+
+	public async generateScriptClick(): Promise<void> {
+		const profile: IGenerateScriptProfile = {
+			databaseName: this.getTargetDatabaseName(),
+			connectionUri: await this.getConnectionUri()
+		};
+
+		if (this.generateScript) {
+			await this.generateScript(this.project, profile);
+		}
```
Comment (Benjin) [inline extensions/sql-database-projects/src/dialogs/deployDatabaseDialog.ts:162]: `await this.generateScript!(...)` (bang)

File: extensions/sql-database-projects/src/controllers/projectController.ts [hunk 9]
```diff
@@ -119,23 +122,77 @@ export class ProjectsController {
...
 		const options: DotNetCommandOptions = {
 			commandTitle: 'Build',
 			workingDirectory: project.projectFolderPath,
 			argument: this.buildHelper.constructBuildArguments(project.projectFilePath, this.buildHelper.extensionBuildDirPath)
 		};
-		await this.netCoreTool.runDotnetCommand(options);
+		try {
+			await this.netCoreTool.runDotnetCommand(options);
+
+			return path.join(project.projectFolderPath, 'bin', 'Debug', `${project.projectFileName}.dacpac`);
```
Comment (Charles-Gagnon) [inline extensions/sql-database-projects/src/controllers/projectController.ts:151]: Will this handle different locales as well? (is the Debug folder always named Debug?)

File: extensions/sql-database-projects/src/controllers/projectController.ts [hunk 10]
```diff
@@ -119,23 +122,77 @@ export class ProjectsController {
...
+		}
+		catch (err) {
+			this.apiWrapper.showErrorMessage(constants.projBuildFailed(utils.getErrorMessage(err)));
+			return;
```
Comment (Charles-Gagnon) [inline extensions/sql-database-projects/src/controllers/projectController.ts:155]: `return undefined;` - better to be explicit with return values.

File: extensions/sql-database-projects/src/controllers/projectController.ts [hunk 11]
```diff
@@ -119,23 +122,77 @@ export class ProjectsController {
...
+	/**
+	 * Builds and deploys a project
+	 * @param treeNode a treeItem in a project's hierarchy, to be used to obtain a Project
+	 */
+	public async deployProject(treeNode: BaseProjectTreeItem): Promise<DeployDatabaseDialog>;
+	/**
+	 * Builds and deploys a project
+	 * @param project Project to be built and deployed
+	 */
+	public async deployProject(project: Project): Promise<DeployDatabaseDialog>;
```
Comment (Charles-Gagnon) [inline extensions/sql-database-projects/src/controllers/projectController.ts:168]: Why do you have these two other signatures? Just the base one is enough with the OR'd type

File: extensions/sql-database-projects/src/controllers/projectController.ts [hunk 12]
```diff
@@ -223,6 +280,21 @@ export class ProjectsController {
 
 	//#region Helper methods
 
+	public getDeployDialog(project: Project): DeployDatabaseDialog {
+		return new DeployDatabaseDialog(this.apiWrapper, project);
+	}
+
+	private static async getService(extensionId: string): Promise<mssql.IDacFxService> {
+		const ext = extensions.getExtension(extensionId);
```
Comment (Charles-Gagnon) [inline extensions/sql-database-projects/src/controllers/projectController.ts:288]: Why are you passing in the extension ID? Do you ever expect a different extension to supply the DacFX service?  Considering you cast to mssql.IExtension that seems unlikely

File: extensions/sql-database-projects/src/controllers/projectController.ts [hunk 13]
```diff
@@ -223,6 +280,21 @@ export class ProjectsController {
...
+
+		if (ext) {
+			return (ext.exports as mssql.IExtension).dacFx;
```
Comment (Charles-Gagnon) [inline extensions/sql-database-projects/src/controllers/projectController.ts:291]: The preferred way is to call ext.activate() and the return value from that is the exported API. That way you're guaranteed that it's been activated.

File: extensions/sql-database-projects/src/controllers/projectController.ts [hunk 14]
```diff
@@ -119,23 +122,77 @@ export class ProjectsController {
...
 		deployDatabaseDialog.openDialog();
+
+		return deployDatabaseDialog;
+	}
+
+	public async executionCallback(project: Project, profile: IDeploymentProfile | IGenerateScriptProfile): Promise<mssql.DacFxResult | undefined> {
+		const dacpacPath = await this.buildProject(project);
+
+		if (!dacpacPath) {
+			return; // buildProject() handles displaying the error
```
Comment (Charles-Gagnon) [inline extensions/sql-database-projects/src/controllers/projectController.ts:185]: Preferred to be explicit with `return undefined;`

File: extensions/mssql/src/dacfx/dacFxService.ts [hunk 15]
```diff
@@ -65,8 +65,8 @@ export class DacFxService implements mssql.IDacFxService {
 		);
 	}
 
-	public deployDacpac(packageFilePath: string, targetDatabaseName: string, upgradeExisting: boolean, ownerUri: string, taskExecutionMode: azdata.TaskExecutionMode): Thenable<mssql.DacFxResult> {
-		const params: contracts.DeployParams = { packageFilePath: packageFilePath, databaseName: targetDatabaseName, upgradeExisting: upgradeExisting, ownerUri: ownerUri, taskExecutionMode: taskExecutionMode };
+	public deployDacpac(packageFilePath: string, targetDatabaseName: string, upgradeExisting: boolean, ownerUri: string, taskExecutionMode: azdata.TaskExecutionMode, sqlCommandVariableValues?: Record<string, string>): Thenable<mssql.DacFxResult> {
+		const params: contracts.DeployParams = { packageFilePath: packageFilePath, databaseName: targetDatabaseName, upgradeExisting: upgradeExisting, sqlCommandVariableValues: sqlCommandVariableValues, ownerUri: ownerUri, taskExecutionMode: taskExecutionMode };
```
Comment (alanrenmsft) [inline extensions/mssql/src/dacfx/dacFxService.ts:69]: hmm, why not make the deployDacpac method's input parameter type to be contracts.DeployParams?

File: extensions/mssql/src/dacfx/dacFxService.ts [hunk 16]
```diff
@@ -76,8 +76,8 @@ export class DacFxService implements mssql.IDacFxService {
 		);
 	}
 
-	public generateDeployScript(packageFilePath: string, targetDatabaseName: string, ownerUri: string, taskExecutionMode: azdata.TaskExecutionMode): Thenable<mssql.DacFxResult> {
-		const params: contracts.GenerateDeployScriptParams = { packageFilePath: packageFilePath, databaseName: targetDatabaseName, ownerUri: ownerUri, taskExecutionMode: taskExecutionMode };
+	public generateDeployScript(packageFilePath: string, targetDatabaseName: string, ownerUri: string, taskExecutionMode: azdata.TaskExecutionMode, sqlCommandVariableValues?: Record<string, string>): Thenable<mssql.DacFxResult> {
```
Comment (alanrenmsft) [inline extensions/mssql/src/dacfx/dacFxService.ts:79]: same as above (simplify the parameter and make its type contracts.GenerateDeployScriptParams)

File: extensions/sql-database-projects/src/common/constants.ts [hunk 17]
```diff
@@ -57,6 +58,7 @@ export const buildDacpacNotFound = localize('buildDacpacNotFound', "Dacpac creat
 export function projectAlreadyOpened(path: string) { return localize('projectAlreadyOpened', "Project '{0}' is already opened.", path); }
 export function projectAlreadyExists(name: string, path: string) { return localize('projectAlreadyExists', "A project named {0} already exists in {1}.", name, path); }
 export function mssqlNotFound(mssqlConfigDir: string) { return localize('mssqlNotFound', "Could not get mssql extension's install location at {0}", mssqlConfigDir); }
+export function projBuildFailed(errorMessage: string) { return localize('projBuildFailed', "Build failed.  Check output pane for more details.  {0}}", errorMessage); }
```
Comment (alanrenmsft) [inline extensions/sql-database-projects/src/common/constants.ts:61]: there are 2 places in the resource string that are using double space, also there is an extra } at the end.

File: extensions/sql-database-projects/src/controllers/projectController.ts [hunk 18]
```diff
@@ -9,18 +9,21 @@ import * as dataSources from '../models/dataSources/dataSources';
 import * as utils from '../common/utils';
```
Comment (alanrenmsft) [inline extensions/sql-database-projects/src/controllers/projectController.ts:9]: please reorder the import statements, there is a vscode extension that does the reordering: vsc-organize-imports

File: extensions/sql-database-projects/src/controllers/projectController.ts (hunk 9 above)
Comment (udeeshagautam) [inline extensions/sql-database-projects/src/controllers/projectController.ts:151]: You should be able to read this path from project File xml (in case someone has custom path)

File: extensions/sql-database-projects/src/dialogs/deployDatabaseDialog.ts (hunk 2 above)
Comment (udeeshagautam) [reply extensions/sql-database-projects/src/dialogs/deployDatabaseDialog.ts:130]: So, if datasource connection is picked - and no password is provided - deploy will try to go through and fail? that might feel broken a bit... might be easier to just open the connection dialog and then take a connection from connection dialog. 
userConnection = await azdata.connection.openConnectionDialog(undefined, connectionProfile);


(schema compare scmp open

File: extensions/sql-database-projects/src/dialogs/deployDatabaseDialog.ts [hunk 19]
```diff
@@ -39,12 +42,12 @@ export class DeployDatabaseDialog {
 		this.initializeDialog();
 		this.dialog.okButton.label = constants.deployDialogOkButtonText;
 		this.dialog.okButton.enabled = false;
-		this.dialog.okButton.onClick(async () => await this.deploy());
+		this.dialog.okButton.onClick(async () => await this.deployClick());
 
 		this.dialog.cancelButton.label = constants.cancelButtonText;
 
 		let generateScriptButton: azdata.window.Button = azdata.window.createButton(constants.generateScriptButtonText);
-		generateScriptButton.onClick(async () => await this.generateScript());
```
Comment (alanrenmsft) [inline extensions/sql-database-projects/src/dialogs/deployDatabaseDialog.ts:47]: the return type of the onClick is disposable, please handle the disposables properly.

File: extensions/mssql/src/dacfx/dacFxService.ts (hunk 15 above)
Comment (Benjin) [reply extensions/mssql/src/dacfx/dacFxService.ts:69]: I think that makes sense; I can make that change in another PR, since it'll have impact across multiple extensions.  Any idea why it was done this way to begin with?

File: extensions/mssql/src/dacfx/dacFxService.ts (hunk 16 above)
Comment (Benjin) [reply extensions/mssql/src/dacfx/dacFxService.ts:79]: (see reply to above comment)

File: extensions/sql-database-projects/src/common/constants.ts (hunk 17 above)
Comment (Benjin) [reply extensions/sql-database-projects/src/common/constants.ts:61]: I was always taught two spaces between sentences?  But I'll begrudgingly change it to one since that appears to be how the rest of ADS is. :)

File: extensions/sql-database-projects/src/controllers/projectController.ts [hunk 20]
```diff
@@ -223,6 +280,30 @@ export class ProjectsController {
...
+		return new DeployDatabaseDialog(this.apiWrapper, project);
+	}
+
+	private static getProjectFromContext(context: Project | BaseProjectTreeItem) {
+		if (context instanceof Project) {
+			return context;
+		}
+
+		if (context.root instanceof ProjectRootTreeItem) {
+			return (context.root as ProjectRootTreeItem).project;
```
Comment (llali) [inline extensions/sql-database-projects/src/controllers/projectController.ts:293]: check for undefined?

File: extensions/sql-database-projects/src/controllers/projectController.ts [hunk 21]
```diff
@@ -223,6 +280,30 @@ export class ProjectsController {
...
+		}
+		else {
+			throw new Error('Unable to establish project context.  Command invoked from unexpected location: ' + context.uri.path);
```
Comment (llali) [inline extensions/sql-database-projects/src/controllers/projectController.ts:296]: add to contants

File: extensions/sql-database-projects/src/models/dataSources/sqlConnectionStringSource.ts [hunk 22]
```diff
@@ -25,6 +25,27 @@ export class SqlConnectionDataSource extends DataSource {
...
+		return this.getSetting(constants.integratedSecuritySetting).toLowerCase() === 'true';
+	}
+
+	public get username(): string {
+		return this.getSetting(constants.userIdSetting);
+	}
+
+	public get password(): string {
+		// TODO: secure password storage
+		return this.getSetting(constants.passwordSetting);
```
Comment (llali) [inline extensions/sql-database-projects/src/models/dataSources/sqlConnectionStringSource.ts:46]: azdata already has APIs to store, load password in secure storage

File: extensions/sql-database-projects/src/controllers/projectController.ts [hunk 23]
```diff
@@ -114,28 +117,82 @@ export class ProjectsController {
...
+	 * @returns path of the built dacpac
+	 */
+	public async buildProject(treeNode: BaseProjectTreeItem): Promise<string>;
+	/**
+	 * Builds a project, producing a dacpac
+	 * @param project Project to be built
+	 * @returns path of the built dacpac
+	 */
+	public async buildProject(project: Project): Promise<string>;
+	public async buildProject(context: Project | BaseProjectTreeItem): Promise<string | undefined> {
```
Comment (udeeshagautam) [inline extensions/sql-database-projects/src/controllers/projectController.ts:137]: are there three of these intentionally?

File: extensions/sql-database-projects/src/controllers/projectController.ts (hunk 9 above)
Comment (udeeshagautam) [reply extensions/sql-database-projects/src/controllers/projectController.ts:151]: Can you open a task/bug on this please so its tracked.

File: extensions/sql-database-projects/src/controllers/projectController.ts (hunk 21 above)
Comment (udeeshagautam) [reply extensions/sql-database-projects/src/controllers/projectController.ts:296]: loc?

File: extensions/sql-database-projects/src/models/dataSources/sqlConnectionStringSource.ts (hunk 22 above)
Comment (udeeshagautam) [reply extensions/sql-database-projects/src/models/dataSources/sqlConnectionStringSource.ts:46]: add a task for this also if you cant take in the PR?

**Comment by udeeshagautam** [comment]
> variable

please test that having sqlcmd vars here but not in sqltoolsservice wouldn't cause any issue in calls - sometimes if calls don't match sqltoolsservice may fail to start correctly)

File: extensions/sql-database-projects/src/models/dataSources/sqlConnectionStringSource.ts (hunk 22 above)
Comment (Benjin) [reply extensions/sql-database-projects/src/models/dataSources/sqlConnectionStringSource.ts:46]: Created #10561

File: extensions/sql-database-projects/src/controllers/projectController.ts (hunk 23 above)
Comment (Benjin) [reply extensions/sql-database-projects/src/controllers/projectController.ts:137]: Yup, it lets me give better docstrings for the individual cases.  I don't love it, but it's how method overloads in typescript work.

**Comment by Benjin** [comment]
It's been running fine without changes in SQL Tool Service; let me know if there are any particular scenarios you'd like me to validate.

File: extensions/sql-database-projects/src/test/projectController.test.ts [hunk 24]
```diff
@@ -27,43 +30,99 @@ describe('ProjectsController: project controller operations', function (): void
...
+			let holler = 'nothing';
+
+			let deployDialog = TypeMoq.Mock.ofType(DeployDatabaseDialog, undefined, undefined, new ApiWrapper(), proj);
+			deployDialog.callBase = true;
+			deployDialog.setup(x => x.getConnectionUri()).returns(async () => 'fake|connection|uri');
+
+			let projController = TypeMoq.Mock.ofType(ProjectsController);
+			projController.callBase = true;
+			projController.setup(x => x.getDeployDialog(TypeMoq.It.isAny())).returns(() => deployDialog.object);
+			projController.setup(x => x.executionCallback(TypeMoq.It.isAny(), TypeMoq.It.is((_): _ is IDeploymentProfile => true))).returns(async () => {
```
Comment (alanrenmsft) [inline extensions/sql-database-projects/src/test/projectController.test.ts:107]: TypeMoq.It.is((_): 

what does this mean?

Comment (Benjin) [reply extensions/sql-database-projects/src/test/projectController.test.ts:107]: It's how you provide a lambda function as a conditional for that argument.  In this case, I'm having the mock perform different actions (setting `holler` to different values) depending on whether the second argument is an `IDeploymentProfile` (this line) or an `IGenerateScriptProfile` (line 112).

I could have also done that switching logic within the mock function itself, but doing it in the mock setup seemed more succinct.

Comment (Benjin) [reply extensions/sql-database-projects/src/test/projectController.test.ts:107]: The `_` indicates to typescript that it's a discard, so it shouldn't complain about me not using that variable in the definition of the function.
